recipes/spirv-tools/all/conanfile.py

```diff
@@ -0,0 +1,86 @@
+from conans import ConanFile, tools, CMake
+import os
+
+
+class SpirvtoolsConan(ConanFile):
+    name = "spirv-tools"
+    homepage = "https://github.com/KhronosGroup/SPIRV-Tools/"
+    description = "SPIRV-Tools "
+    topics = ("conan", "spirv", "spirv-v", "vulkan", "opengl", "opencl", "hlsl", "khronos")
+    url = "https://github.com/conan-io/conan-center-index"
+    settings = "os", "compiler", "arch", "build_type"
+    exports_sources = ["CMakeLists.txt"]
+    license = "Apache-2.0"
+    generators = "cmake"
+
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "c_only": [True, False]
+    }
+    default_options = {
+        "shared": True,
+        "fPIC": True,
+        "c_only": False
```

Comment:
I don't think this option should be provided. `SPIRV-Tools-shared` should be completly removed.

With the current recipe:
- if `spirv-tools:shared=True`: `SPIRV-Tools-shared` and `SPIRV-Tools` seems to be the same shared lib. But only `SPIRV-Tools` is linked into `SPIRV-Tools-reduce` and `SPIRV-Tools-opt`

Reply:
I think that shared is simply not supported in SPIRV-tools at least on Windows, because others libs than `SPIRV-Tools-shared` (`SPIRV-Tools`, `SPIRV-Tools-reduce`, `SPIRV-Tools-link`, `SPIRV-Tools-opt`) don't export their symbols.
Therefore, I would recommend to patch CMakeLists in order to completly remove any reference to `SPIRV-Tools-shared`, and raise a `ConanInvalidConfiguration` if shared (maybe only for Windows).

Reply:
The `spirv-tools-shared.so` confused me as well. You would link only to this library if you were building a C-only application. `spirv-tools-shared.so` contains all the code from the other libraries, but only exposes the C interface. 

Running `patchelf --print-needed spirv-tools-shared.so` shows that it does not link to any of the other built shared libraries.

What may need to be done is delete the other shared objects, keeping only `spirv-tools-shared.so` when `c_only=true`.

The C-Interface cannot be built as a static library. So I'm not entirely sure what to do about `shared=false` when `c_only=true`. Maybe throw the `ConanInvalidConfiguration` error saying `"the C-only interface MUST be built as a shared library" ?

I don't use windows, does `test_package.cpp` compile and link on windows when `shared=True` and `c_only=False`?

Reply:
You're right, `spirv-tools-shared` is not the combination of all the others. It looks like it's just `spirv-tools.so` without the C++ interface.

https://github.com/KhronosGroup/SPIRV-Tools/blob/8dd174809fb86498e5870276be7c84a8bf4c8254/source/CMakeLists.txt#L350

https://github.com/KhronosGroup/SPIRV-Tools/blob/8dd174809fb86498e5870276be7c84a8bf4c8254/source/CMakeLists.txt#L363

Reply:
Here is the summary of the two options I'm thinking of.

1. Do not allow spirv-tools-shared.so to be built. C-only users are on their own. This allows the rest of the libraries to be built in shared/static mode

2. If `c_only=true`,  Keep only spirv-tools-shared.so and delete the other libraries. C-only users will have to live without the use cases of the other libraries. Throw a `ConanInvalidConfiguration` error if this `c_only=True` and `shared=False`. This allows the C++ interface to be built in shared/static mode, and the C-only interface to only be built in `shared` mode.

I'm really surprised at the Khronos group's CMake skills. For an organisation of their reputation, they're CMake skills are very lacking, lol. I've had such a hard time compiling any of their libraries

Reply:
For me this option `c_only` doesn't make sense. In your recipe it doesn't change the binaries produced, it only hides some libs in cpp_info.
All C++ libraries with C API would have this option `c_only` if you want, but there is already an option for that: it's `shared=True`. Moreover they can link the static libs, but they must also link libstdc++.

I don't know what happened in Jenkins with this recipe, but I'm pretty sure that the build breaks at least with `shared=True`

Here is a way to handle all this mess:
1/ Only build specifics targets (`cmake.build(target="foo")`):
- if `shared=True`: `SPIRV-Tools-shared`
- if `shared=False`: `SPIRV-Tools-reduce`, `SPIRV-Tools-link`, `SPIRV-Tools-opt`, `SPIRV-Tools` + executables (`spirv-as`, `spirv-dis`, etc...)

2/ if `shared=True`, as a second step, you will have to create another build directory with the same configuration as `shared=False` and build executables only (it will automacally build static libs since they are dependencies).

3/ Do not use `cmake.install()` because it will build what you didn't build previously due to how it is done in the library. Therefore you will have to copy by hand proper include and lib files

Other solution: patch CMakeLists, there is a lot of work I guess.

Reply:
For 1/, I see some potential issues.
- if a user is building a C-only application, they would set `shared=True`, and everything would work.
- But if a user is building a C++ application, and they set `shared=True`, then conan will link with `SPIRV-Tools-shared`, which is not what they want, they would want the other libraries because they have the C++ interface.

We would still need an option to differentiate between `c/c++` and `shared/static`, only  `c/shared`, `c++/shared` and `c++/static` combinations would be allowed. `c/static` would not be allowed. If you are not happy with having a "invalid option" then we'd have to remove the `c/shared` option completely and only allow C++ usage.

The if the user is linking to to spirv-tools-shared, they will still need libstdc++ because spirv-tools-shared requires it. It only needs `libspirv.h` instead of `libspirv.hpp` to use.

/2, I don't see a point in having a second build directory, the executables are built using C++, and therefore will always link to the C++ libs, either static/shared. 

3/ What's wrong with using cmake.install() and then just deleting the headers/libraries that aren't part of that combination of options?

Reply:
There is another solution:
- whatever the value of `shared`, we force `BUILD_SHARED_LIBS OFF`
- if shared=True, we remove `SPIRV-Tools` 
- if shared=False, we remove `SPIRV-Tools-shared`

Maybe it's the best solution with the current state of this library.

Reply:
But that still leaves the problem that if we have `shared=True` we cannot use C++.

The issue isn't the `shared/static` the issue is whether we want the C library

Reply:
In the current state, the best option is to ditch the C library (ie: remove SPIRV-Tools-shared)

Reply:
Tested, and yes it can't work because others libs use C++ API of `SPIRV-Tools` (which only exports C symbols).

Reply:
Now I understand why you added `c_only` option.

Other suggestion:
- if `shared=False` and `c_only=False`: build as usual and remove `SPIRV-Tools-shared` (also in package folder).
- if `shared=False` and `c_only=True`: build as usual and remove all libs from package folder except `SPIRV-Tools`, and all public headers in include folder except `libspirv.h`
- if `shared=True` and `c_only=False`: raise ConanInvalidConfiguration
- if `shared=True` and `c_only=True`: force `BUILD_SHARED_LIBS OFF`, build and remove all libs from package folder except `SPIRV-Tools-shared`, and all public headers in include folder except `libspirv.h`. In this case, add `SPIRV_TOOLS_SHAREDLIB` in `cpp_info.defines`

`c_only=False` should still be default value because `shaderc`, `glslang` and `Vulkan-ValidationLayers` use C++ API (at least `SPIRV` in `glslang` use C++ API since it may link `SPIRV-Tools-opt`).

Reply:
I think you have one the configs backwards. I think it should be:

* if `shared=False` and `c_only=False` build as usual and remove `spirv-tools-shared` (spirv-tools-shared is C-only, not needed by C++)

* if `shared=False` and `c_only=True` raise `conanInvalidConfig` (cannot produce a C-only static lib)

* if `shared=True` and `c_only=False` build as usual with `BUILD_SHARED_LIBS  ON`, remove `spirv-tools-shared` (not needed by C++)

* if `shared=True` and `c_only=True` build as usual, delete everything except `spirv-tools-shared`, all the `.hpp` header files, add `SPIRV_TOOLS_SHAREDLIB` definition for windows. (SPIRV_TOOLS_SHAREDLIB looks like it's only used for `spirv-tools-shared` and not the other ones, see: https://github.com/KhronosGroup/SPIRV-Tools/blob/61b7de3c39f01a0eeb717f444c86990547752e26/source/CMakeLists.txt#L373 )



(hopefully I got that right)

Reply:
Sure you can't produce a C only static lib, but you can compile with C API, and link with libstdc++ (I understand `c_only` as `c_api_only`).
You can't use BUILD_SHARED_LIBS ON in this lib with Visual Studio as is, because a lot of targets will fail to link against others. Maybe you can try to use CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS in this case for `shared=True` and `c_only=False`. But you don't want to use CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS  with `shared=True` and `c_only=True` because it will lead to less optimal shared lib.

Reply:
Yeah, you're right it should probably be `c_api_only`.  I'll make some changes and do a new push and you can look at it.

Reply:
I've created a testing repo here: https://github.com/SpaceIm/conan-spirv-tools
It seems to work, except for MinGW.

You can see several suggestions implemented:
- force `BUILD_SHARED_LIBS ON` if shared and c_api
- force `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` if shared, not c_api and Visual Studio
- remove files in package folder depending on options
- add `SPIRV_TOOLS_SHAREDLIB` in `cpp_info.defines` if shared and c_api_only
- add `rt` to system_libs on Linux
- remove `CMAKE_POSITION_INDEPENDENT_CODE` in CMakeLists
- always build C test_package
- build C++ test_package only if `c_api_only` is False

Also:
- Disable SPIRV_CHECK_CONTEXT

Reply:
Looking closely to this library and discussions in issues, it appears that there shouldn't be a `c_api_only` (or `c_only`) option.
They try to fix the shared libs, but it will be very hard because most of them and also their executables also depends on private headers !

For this version, I would:
- patch CMakeLists.txt to remove any reference to SPIRV-Tools-shared
- remove `c_only` option
- either raise ConanInvalidConfiguration for shared and Visual Studio, or use CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS (which is bad practice but there is no easy solution without this, I've tried to fix the libs but it's really hard).

Reply:
Okay I can do that. Is there a reason to explicitly patch the CMakeLists.txt file to remove SPIRV-Tools-shared? Can't we just delete the file after building/before packaging?

Reply:
you're probably right, it should work. By the way, you should only add one version (2020.2 or 2019.2 as you wish. At some point, I will require 2020.2 for recent versions of Vulkan-ValidationLayers since 2019.2 is not compatible with them).

Reply:
you can look at:
https://github.com/SpaceIm/conan-spirv-tools/blob/testing/2019.2/conanfile.py
https://github.com/SpaceIm/conan-spirv-tools/blob/testing/2019.2/test_package/CMakeLists.txt
https://github.com/SpaceIm/conan-spirv-tools/blob/testing/2019.2/test_package/conanfile.py

Reply:
@SpaceIm , I made the changes you suggested. 

I probably shouldn't have included 2020.1 since it's not a stable release. The reason it was there was actually because the glslang package that I'm working on needs a newer version of the headers to enable HSLSL. I'll remove 2020.1 until that branch has released a stable version.

I'll be pushing a glslang conan package after this spirv-tools is merged, but it wont have HLSL support until a newer version of spirv-tools is released in conan